### PR TITLE
delete DSA support

### PR DESCRIPTION
**Pull request by xiaoge1001**
## Proposed Commit Message
OpenSSH 7.0 and greater similarly disable the ssh-dss (DSA) public key algorithm. It too is weak and we recommend against its use. 
refer:https://www.openssh.com/legacy.html

However, cloud-init still generates the DSA key.  It is recommended that cloud-init disable dsa

### CODE REVIEW

**Comment by xiaoge1001** [comment]
Does anyone care about this?

**Comment by blackboxsw** [comment]
Thank you @xiaoge1001 for the proposal here. It is very broad, but I agree it's probably ideal to drop from upstream cloud-init. Brett already added the security label to indicate we'll need security review on this approach to ensure we understand implication for any other potential downstream offerings (such as stable releases of distributions that still grab latest cloud-init). To determine whether or not this general solution is considered viable on some of those stable downstreams. Regardless, we'll get a review on this from upstream perspective and see whether we need to adapt your approach to still leave room for other distributions to continue to include the legacy dsa keys.

**Comment by xiaoge1001** [comment]
@blackboxsw I am pleased to receive your reply and look forward to the results as soon as possible.

**Comment by xiaoge1001** [comment]
I have a question, why delete the existing /etc/ssh/ssh_host_*_key*?

https://github.com/canonical/cloud-init/blob/main/cloudinit/config/cc_ssh.py#L202

**Comment by tobhe** [comment]
This diff looks good as far as I'm concerned. DSA really shouldn't be used in 2023. OpenSSH deprecated support by default in 2015, so I think there isn't a huge risk in dropping them either. I think almost all of the ecosystem has moved on to (bigger) rsa keys or elliptic curves.

File: cloudinit/config/cc_phone_home.py
Comment (blackboxsw) [inline cloudinit/config/cc_phone_home.py:24]: Security review heads up: team behavior changes in stable releases ( 20.04, 22.04) here will only trigger warnings and skip `pub_key_dsa` if provided  in ser-data or system /etc/cloud/cloud-config.*. So images or users how want to report DSA pub_keys within phone_home directives would not have an option to phone home with DSA keys anymore if we backport this change to stable releases. This is probably reasonable behavior for any stable release which still gets updates from upstream as we want to ensure a base-level of security continues to improve. Those who don't want security-related updates should probably during off update pockets such as `jammy-updates`

File: cloudinit/config/cc_ssh.py
Comment (blackboxsw) [inline cloudinit/config/cc_ssh.py:144]: security review: Do we want to fall back here if the host itself has openssl is built with `enable-weak-ssl-ciphers`? Could we check this with `openssl ciphers -s DSA` ? Or are we okay stating folks are off-plan on stable Ubuntu 20.04 if they still rely on DSA . and they'd need to generate their own host-keys with either supplemental `#cloud-config`  via either [write_files](https://cloudinit.readthedocs.io/en/latest/reference/modules.html#write-files) or [runcmd](https://cloudinit.readthedocs.io/en/latest/reference/modules.html#runcmd) customization

**Review by blackboxsw — COMMENTED** [review]
Just a couple of thoughts/questions here for our security team. Thanks @tobhe

File: cloudinit/config/cc_ssh.py
Comment (blackboxsw) [inline cloudinit/config/cc_ssh.py:46]: @holmanb do you think it's worth us documenting here that DSA host key generation dropped in 23.4, or shall we prefer downstreams to rely on the [upstream project Changelog (and release changelog) here](https://github.com/canonical/cloud-init/blob/main/ChangeLog)

File: cloudinit/config/cc_phone_home.py
Comment (tobhe) [reply cloudinit/config/cc_phone_home.py:24]: Ack, I think that's fine.  The security benefit is certainly bigger than the loss. If anyone even notices they can just switch their scripts to use RSA.

File: cloudinit/config/cc_ssh.py
Comment (tobhe) [reply cloudinit/config/cc_ssh.py:144]: I don't think that a fallback is necessary. Ubuntu has disabled support for ssh-dss in 16.04, so there aren't a whole lot of clients around that would work with this key and those that do probably shouldn't have network access.

**Comment by tobhe** [comment]
Github won't let me highlight those but there are a few more occurrences of ssh-dss that can be deleted. grep says:
```
$ grep -r ssh-dss
tests/unittests/test_ssh_util.py:    "ssh-dss-cert-v01@openssh.com": (
tests/unittests/test_ssh_util.py:    "ssh-dss": (
tests/integration_tests/modules/test_keys_to_console.py:ssh_fp_console_blacklist: [ssh-dss, ssh-dsa, ecdsa-sha2-nistp256]
tests/integration_tests/modules/test_keys_to_console.py:ssh_key_console_blacklist: [ssh-dss, ssh-dsa, ecdsa-sha2-nistp256]
tests/integration_tests/modules/test_keys_to_console.py:ssh_fp_console_blacklist: [ssh-dsa, ssh-ecdsa, ssh-ed25519, ssh-rsa, ssh-dss, ecdsa-sha2-nistp256]
cloudinit/config/cc_keys_to_console.py:        " ``ssh-dss`` host keys are not written to console."
cloudinit/config/cc_keys_to_console.py:            ssh_key_console_blacklist: [dsa, ssh-dss]
cloudinit/config/cc_keys_to_console.py:        cfg, "ssh_key_console_blacklist", ["ssh-dss"]
cloudinit/config/cc_ssh.py:    - ssh-dss-cert-v01@openssh.com
cloudinit/config/cc_ssh.py:    - ssh-dss
cloudinit/config/schemas/schema-cloud-config-v1.json:            "ssh-dss"
cloudinit/ssh_util.py:    "ssh-dss-cert-v01@openssh.com",
cloudinit/ssh_util.py:    "ssh-dss",
doc/examples/cloud-config.txt:# ssh_key_console_blacklist: [ssh-dss]
```

File: cloudinit/config/cc_ssh.py
Comment (holmanb) [reply cloudinit/config/cc_ssh.py:46]: I lean towards the later. If a user does actually exist, and upgrading cloud-init breaks their configuration, then looking at the docs will tell them that no such key exists. Given the unlikelyhood of this still being used I doubt it is worth it, and as you pointed out - the changelogs will document this change, as will the commit history and the change in docs between versions.

**Comment by xiaoge1001** [comment]
> Github won't let me highlight those but there are a few more occurrences of ssh-dss that can be deleted. grep says:
> 
> ```
> $ grep -r ssh-dss
> tests/unittests/test_ssh_util.py:    "ssh-dss-cert-v01@openssh.com": (
> tests/unittests/test_ssh_util.py:    "ssh-dss": (
> tests/integration_tests/modules/test_keys_to_console.py:ssh_fp_console_blacklist: [ssh-dss, ssh-dsa, ecdsa-sha2-nistp256]
> tests/integration_tests/modules/test_keys_to_console.py:ssh_key_console_blacklist: [ssh-dss, ssh-dsa, ecdsa-sha2-nistp256]
> tests/integration_tests/modules/test_keys_to_console.py:ssh_fp_console_blacklist: [ssh-dsa, ssh-ecdsa, ssh-ed25519, ssh-rsa, ssh-dss, ecdsa-sha2-nistp256]
> cloudinit/config/cc_keys_to_console.py:        " ``ssh-dss`` host keys are not written to console."
> cloudinit/config/cc_keys_to_console.py:            ssh_key_console_blacklist: [dsa, ssh-dss]
> cloudinit/config/cc_keys_to_console.py:        cfg, "ssh_key_console_blacklist", ["ssh-dss"]
> cloudinit/config/cc_ssh.py:    - ssh-dss-cert-v01@openssh.com
> cloudinit/config/cc_ssh.py:    - ssh-dss
> cloudinit/config/schemas/schema-cloud-config-v1.json:            "ssh-dss"
> cloudinit/ssh_util.py:    "ssh-dss-cert-v01@openssh.com",
> cloudinit/ssh_util.py:    "ssh-dss",
> doc/examples/cloud-config.txt:# ssh_key_console_blacklist: [ssh-dss]
> ```

Thanks for pointing out, I didn't notice it before.

**Comment by xiaoge1001** [comment]
There seems to be some code that needs to be changed if we want to remove the DSA, not just the ssh-dss mentioned by @tobhe .

**Comment by xiaoge1001** [comment]
Does anyone care about this?

**Comment by holmanb** [comment]
@xiaoge1001 The state of this PR is unclear. Your previous comment indicates that more work is to be done with ssh-dss and other things. Since more work is required, it seems that this PR is incomplete. I'd prefer to see the remaining cleanup code land in this PR, rather than leaving cleanup work to a separate commit.

Are you planning to do this work?

**Comment by xiaoge1001** [comment]
@holmanb Sorry, No plans recently.

**Comment by blackboxsw** [comment]
> @holmanb Sorry, No plans recently.

Thank you for the reply @xiaoge1001 I'll either provide a supplemental commit for this changeset to this PR to get it across the line for security recommendations or we will provide a separate PR right after this to sort the ssh-dss references and defaults.

**Comment by xiaoge1001** [comment]
> > @holmanb Sorry, No plans recently.
> 
> Thank you for the reply @xiaoge1001 I'll either provide a supplemental commit for this changeset to this PR to get it across the line for security recommendations or we will provide a separate PR right after this to sort the ssh-dss references and defaults.

Okay, thanks. If you intend to will provide a separate PR, you can close the current PR.

**Review by blackboxsw — APPROVED** [review]
I'm +1 now that ssh-dss references are dropped and `ssh_key_console_blacklist` no longer defines stale dsa or ssh-dss key types. I have added only a supplemental commit to @xiaoge1001's approach to drop remaining ssh-dss references throughout docs, comments and config defaults. Let's get one more set of eyes on my supplemental commit before we land it. Many thanks

**Review by TheRealFalcon — COMMENTED** [review]
A few more references we could remove:
```diff
diff --git a/tests/data/merge_sources/expected9.yaml b/tests/data/merge_sources/expected9.yaml
index 00f91ca06..912276b46 100644
--- a/tests/data/merge_sources/expected9.yaml
+++ b/tests/data/merge_sources/expected9.yaml
@@ -2,4 +2,4 @@
 
 phone_home:
  url: http://my.example.com/$INSTANCE_ID/$BLAH_BLAH
- post: [ pub_key_dsa, pub_key_rsa, pub_key_ecdsa, instance_id ]
+ post: [ pub_key_rsa, pub_key_ecdsa, instance_id ]
diff --git a/tests/data/merge_sources/source9-1.yaml b/tests/data/merge_sources/source9-1.yaml
index 0b102ba6c..d8dc9f174 100644
--- a/tests/data/merge_sources/source9-1.yaml
+++ b/tests/data/merge_sources/source9-1.yaml
@@ -2,4 +2,4 @@
 
 phone_home:
  url: http://my.example.com/$INSTANCE_ID/
- post: [ pub_key_dsa, pub_key_rsa, pub_key_ecdsa, instance_id ]
+ post: [ pub_key_rsa, pub_key_ecdsa, instance_id ]
diff --git a/tests/unittests/config/test_cc_ssh.py b/tests/unittests/config/test_cc_ssh.py
index 747705deb..f59160c16 100644
--- a/tests/unittests/config/test_cc_ssh.py
+++ b/tests/unittests/config/test_cc_ssh.py
@@ -20,9 +20,6 @@ from tests.unittests.util import get_cloud
 LOG = logging.getLogger(__name__)
 
 MODPATH = "cloudinit.config.cc_ssh."
-KEY_NAMES_NO_DSA = [
-    name for name in cc_ssh.GENERATE_KEY_NAMES if name not in "dsa"
-]
 
 
 @pytest.fixture(scope="function")
@@ -226,10 +223,10 @@ class TestHandleSsh:
     @pytest.mark.parametrize(
         "cfg, expected_key_types",
         [
-            pytest.param({}, KEY_NAMES_NO_DSA, id="default"),
+            pytest.param({}, cc_ssh.GENERATE_KEY_NAMES, id="default"),
             pytest.param(
                 {"ssh_publish_hostkeys": {"enabled": True}},
-                KEY_NAMES_NO_DSA,
+                cc_ssh.GENERATE_KEY_NAMES,
                 id="config_enable",
             ),
             pytest.param(
diff --git a/tools/mock-meta.py b/tools/mock-meta.py
index 15d8c4060..a52536165 100755
--- a/tools/mock-meta.py
+++ b/tools/mock-meta.py
@@ -179,8 +179,6 @@ def get_ssh_keys():
 
     # Nice helper to add in the 'running' users key (if they have one)
     key_pth = os.path.expanduser("~/.ssh/id_rsa.pub")
-    if not os.path.isfile(key_pth):
-        key_pth = os.path.expanduser("~/.ssh/id_dsa.pub")
 
     if os.path.isfile(key_pth):
         with open(key_pth, "rb") as fh:

```

**Comment by blackboxsw** [comment]
> A few more references we could remove:
> 
> ```diff
> diff --git a/tests/data/merge_sources/expected9.yaml b/tests/data/merge_sources/expected9.yaml
> index 00f91ca06..912276b46 100644
> --- a/tests/data/merge_sources/expected9.yaml
> +++ b/tests/data/merge_sources/expected9.yaml
> @@ -2,4 +2,4 @@
>  
>  phone_home:
>   url: http://my.example.com/$INSTANCE_ID/$BLAH_BLAH
> - post: [ pub_key_dsa, pub_key_rsa, pub_key_ecdsa, instance_id ]
> + post: [ pub_key_rsa, pub_key_ecdsa, instance_id ]
> diff --git a/tests/data/merge_sources/source9-1.yaml b/tests/data/merge_sources/source9-1.yaml
> index 0b102ba6c..d8dc9f174 100644
> --- a/tests/data/merge_sources/source9-1.yaml
> +++ b/tests/data/merge_sources/source9-1.yaml
> @@ -2,4 +2,4 @@
>  
>  phone_home:
>   url: http://my.example.com/$INSTANCE_ID/
> - post: [ pub_key_dsa, pub_key_rsa, pub_key_ecdsa, instance_id ]
> + post: [ pub_key_rsa, pub_key_ecdsa, instance_id ]
> diff --git a/tests/unittests/config/test_cc_ssh.py b/tests/unittests/config/test_cc_ssh.py
> index 747705deb..f59160c16 100644
> --- a/tests/unittests/config/test_cc_ssh.py
> +++ b/tests/unittests/config/test_cc_ssh.py
> @@ -20,9 +20,6 @@ from tests.unittests.util import get_cloud
>  LOG = logging.getLogger(__name__)
>  
>  MODPATH = "cloudinit.config.cc_ssh."
> -KEY_NAMES_NO_DSA = [
> -    name for name in cc_ssh.GENERATE_KEY_NAMES if name not in "dsa"
> -]
>  
>  
>  @pytest.fixture(scope="function")
> @@ -226,10 +223,10 @@ class TestHandleSsh:
>      @pytest.mark.parametrize(
>          "cfg, expected_key_types",
>          [
> -            pytest.param({}, KEY_NAMES_NO_DSA, id="default"),
> +            pytest.param({}, cc_ssh.GENERATE_KEY_NAMES, id="default"),
>              pytest.param(
>                  {"ssh_publish_hostkeys": {"enabled": True}},
> -                KEY_NAMES_NO_DSA,
> +                cc_ssh.GENERATE_KEY_NAMES,
>                  id="config_enable",
>              ),
>              pytest.param(
> diff --git a/tools/mock-meta.py b/tools/mock-meta.py
> index 15d8c4060..a52536165 100755
> --- a/tools/mock-meta.py
> +++ b/tools/mock-meta.py
> @@ -179,8 +179,6 @@ def get_ssh_keys():
>  
>      # Nice helper to add in the 'running' users key (if they have one)
>      key_pth = os.path.expanduser("~/.ssh/id_rsa.pub")
> -    if not os.path.isfile(key_pth):
> -        key_pth = os.path.expanduser("~/.ssh/id_dsa.pub")
>  
>      if os.path.isfile(key_pth):
>          with open(key_pth, "rb") as fh:
> ```


Thanks @TheRealFalcon taking your suggestion and found one more irrelevant reference
```
--- a/tests/unittests/config/test_cc_ssh.py
+++ b/tests/unittests/config/test_cc_ssh.py
@@ -131,7 +128,7 @@ class TestHandleSsh:
                 mock.call("/etc/ssh/ssh_host_ed25519_key"),
             ]
         else:
-            # Enabled fips doesn't generate dsa or ed25519
+            # Enabled fips doesn't generate ed25519
             expected_calls = [
                 mock.call("/etc/ssh/ssh_host_rsa_key"),
                 mock.call("/etc/ssh/ssh_host_ecdsa_key"),
@@ -226,10 +223,10 @@ class TestHandleSsh:
```

**Review by TheRealFalcon — APPROVED** [review]
Mypy is angry now, but +1 assuming that gets fixed

**Pull request merged**
